### PR TITLE
fix system_program blind indexing

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 system program blindly indexes into keyed_accounts

 #### Summary of Changes
 check the keyed_accounts length and bail on mismatch

Fixes #

### CODE REVIEW

File: runtime/src/system_instruction_processor.rs [hunk 1]
```diff
@@ -96,28 +133,29 @@ pub fn process_instruction(
...
                 program_id,
-            } => create_system_account(keyed_accounts, lamports, space, &program_id),
-            SystemInstruction::Assign { program_id } => {
-                if !system_program::check_id(&keyed_accounts[FROM_ACCOUNT_INDEX].account.owner) {
-                    return Err(InstructionError::IncorrectProgramId);
-                }
-                assign_account_to_program(keyed_accounts, &program_id)
-            }
-            SystemInstruction::Transfer { lamports } => transfer_lamports(keyed_accounts, lamports),
+            } => with_keyed_accounts!(
```
Comment (rob-solana) [inline runtime/src/system_instruction_processor.rs]: @garious thanks for the pattern matching tricks, using it to good effect here?

**Comment by garious** [comment]
I like the direction a LOT, but hold off on merging this tonight. If we're going to do this everywhere, I'd like to spend a bit more time on it, and make it perfect. I'm thinking `into_iter()` and `next()` might clean things up even more.

**Comment by garious** [comment]
@rob-solana, here's what I had in mind: #6309

**Comment by rob-solana** [comment]
cool.  this is a first pass at stupid Rust tricks.  My need for optional args in stake means this needs more development.

**Comment by rob-solana** [comment]
@garious you might also want to have a look at @CriesofCarrots PR #6314  in stakes, where the parameters *and* the array of keyed accounts is needed for many of stake program's methods...

**Comment by rob-solana** [comment]
abandoning because @garious doesn't like stupid Rust tricks ;)

**Pull request closed without merging**